### PR TITLE
feat: Add rotation to iOS (the sequel)

### PR DESCRIPTION
**Pull request by juanmartin8a**
Continuation of #540. Added a new variable to the `InputImageMetadata` class, `cameraLensDirection`, it is used only on iOS to determine if the image is mirrored (for example when using the front camera) or not.

### CODE REVIEW

**Comment by juanmartin8a** [comment]
The example must change for this to work, sadly I am unable to run the example app on my computer. On android things can stay the same however on iOS the rotation compensation must be calculated based on the orientation of the device. Like this: 
`int getRotationCompensation(int orientation, int sensorOrientation, CameraLensDirection cameraLensDirection) {
  if (Platform.isAndroid) {
    return cameraLensDirection == CameraLensDirection.front 
      ? (sensorOrientation + orientation) % 360
      : (sensorOrientation - orientation + 360) % 360;
  }

  return (360 - orientation) % 360;
}`. Orientation can be calculated using the [native_device_orientation](https://pub.dev/packages/native_device_orientation) package.

IOS does not require different coordinate translation when the device orientation changes, it can be done with just `x * canvasSize.width / imageSize.width;` for x, and `y * canvasSize.height / imageSize.height;` for y. This applies for all orientations.

This works when the camera is locked to portrait up in iOS. If the camera is not being locked, then a rotation of 0 in the InputImage will do.

Finally, the InputImage must be called like this now, 
`
InputImage inputImage = InputImage.fromBytes(
            bytes: image.planes[0].bytes, 
            metadata: InputImageMetadata(
              size: Size(image.width.toDouble(), image.height.toDouble()), 
              rotation: InputImageRotationValue.fromRawValue(rotation)!, 
              format: InputImageFormatValue.fromRawValue(image.format.raw)!, 
              bytesPerRow: image.planes[0].bytesPerRow,
              cameraLensDirection: InputImageCameraLensDirectionValue.fromRawValue(InputImageCameraLensDirection.back.rawValue)!
              )
            );
`. In Android the cameraLensRotation is not used so any value will output the same result. :)

**Comment by fbernaly** [comment]
I need you to test with the example app and edit it if needed before merging this.

File: packages/google_mlkit_commons/ios/Classes/MLKVisionImage+FlutterPlugin.m [hunk 1]
```diff
@@ -6,9 +6,19 @@ @implementation MLKVisionImage(FlutterPlugin)
 + (MLKVisionImage *)visionImageFromData:(NSDictionary *)imageData {
     NSString *imageType = imageData[@"type"];
     if ([@"file" isEqualToString:imageType]) {
-        return [self filePathToVisionImage:imageData[@"path"]];
+        int rotation = [imageData[@"metadata"][@"rotation"] intValue];
```
Comment (fbernaly) [inline packages/google_mlkit_commons/ios/Classes/MLKVisionImage+FlutterPlugin.m:9]: This is not needed. When the image is a file, camera lens is not used.
Also, check what it is happening inside `filePathToVisionImage` : https://github.com/flutter-ml/google_ml_kit_flutter/blob/develop/packages/google_mlkit_commons/ios/Classes/MLKVisionImage%2BFlutterPlugin.m#L20-L25

**Review by fbernaly — CHANGES_REQUESTED** [review]

File: packages/google_mlkit_commons/ios/Classes/MLKVisionImage+FlutterPlugin.m (hunk 1 above)
Comment (fbernaly) [reply packages/google_mlkit_commons/ios/Classes/MLKVisionImage+FlutterPlugin.m:9]: Orientation is already set in that method for files.

**Comment by juanmartin8a** [comment]
> I need you to test with the example app and edit it if needed before merging this.

It doesn't run in my computer (m1 macbook pro), I tried many times. The alternative is my parents computer which will probably melt if I tried to run it haha.

File: packages/google_mlkit_commons/ios/Classes/MLKVisionImage+FlutterPlugin.m (hunk 1 above)
Comment (juanmartin8a) [reply packages/google_mlkit_commons/ios/Classes/MLKVisionImage+FlutterPlugin.m]: Sorry, I didn't noticed that. I already made a new commit with those changes.

**Comment by fbernaly** [comment]
I will give it a try

**Comment by juanmartin8a** [comment]
Thank you!

> I will give it a try

Thank you!

**Comment by fbernaly** [comment]
@juanmartin8a : Thanks for using this plugin and your contribution.

I merged your PR into this branch:  [feature/rotation](https://github.com/flutter-ml/google_ml_kit_flutter/tree/feature/rotation) 
and I was able to compile it and play with it, but unfortunately the `imageOrientationFromRotation:cameraLensDirection:` method  in this [file](https://github.com/flutter-ml/google_ml_kit_flutter/blob/feature/rotation/packages/google_mlkit_commons/ios/Classes/MLKVisionImage%2BFlutterPlugin.m#L74-L89) is breaking things for our example app.

Since we are not setting `MLKVisionImage.orientation`,  by default the display orientation of the image is `.up`:

Screenshot 2024-04-15 at 4 49 01 PM

and with your method you are introducing different rotations that are not handled in the example app and as a result it is not detecting any objects/faces/etc because the rotation is messed up.

I believe most of the devs consuming this plugin are doing the image rotation as we have it in the example app and explained in the README:
https://github.com/flutter-ml/google_ml_kit_flutter/tree/feature/rotation/packages/google_mlkit_commons#creating-an-inputimage

If we release your changes we will break a lot of projects once they pull the latest version with your changes.

If you were to fix the example app to work with your changes we will merge them and release them. In the meantime we will keep them for a while in the [feature/rotation branch](https://github.com/flutter-ml/google_ml_kit_flutter/tree/feature/rotation).

Let me know if you are not interested in contributing more so we could archive this branch.

Let me know your comments.

**Comment by fbernaly** [comment]
I have an M1 as well, why you can not run it?

**Comment by juanmartin8a** [comment]
It was something about the compile size being too big and it being a common problem in the ARM architecture. I'm going to try again today and will let you know :)

**Comment by fbernaly** [comment]
maybe it was this issue:

https://forums.developer.apple.com/forums/thread/738275

got that one when I ran the app in Xcode 15, but I fixed it: https://github.com/flutter-ml/google_ml_kit_flutter/commit/2d3b4f81e69393fda0455995fb7cbd7a6329d371

Pull the latest code from feature/rotation

**Comment by juanmartin8a** [comment]
Hi! Sorry for the late reply. The example works! I pulled from the develop branch and will be working on making the rotation backwards compatible.

**Pull request merged**